File: app/src/main/java/de/dariatech/softphone/Verzeichnis.kt
package de.dariatech.softphone

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Handler
import android.os.Looper
import java.io.File
import java.util.concurrent.Executors

/**
 * Das Verzeichnis im Gerät: wer ruft da an, und wie sieht er aus?
 *
 * DER AUFTRAG (Inhaber, 05.09.2026): „…und wenn man sich gegenseitig
 * anruft, dass das Foto von dem Anruf angezeigt wird."
 *
 * DIE RÜCKWÄRTSSUCHE MACHT DIE APP SELBST. So steht es in
 * `pbx/docs/CLIENT-API.md`, und der Grund ist hart: Sie muss bei JEDEM
 * Anruf funktionieren, auch ohne Netz. Wer beim Klingeln erst den Server
 * fragt, zeigt den Namen nach dem dritten Klingeln – oder gar nicht.
 *
 * ZUM VERGLEICHEN WIRD NORMALISIERT, ZUM ANZEIGEN NICHT. „+49 831 555 12",
 * „0049831555 12" und „0831 55512" sind dieselbe Nummer; angezeigt wird
 * die Schreibweise, die der Mensch selbst gewählt hat – sie ist die, die
 * er wiedererkennt.
 *
 * BILDER LIEGEN IM GERÄT, benannt nach ihrer MARKE. Ändert sich das
 * Bild, ändert sich der Name – und nur dann wird geladen. Ein Telefon
 * fragt bei jedem Klingeln; ohne das lüde es dasselbe Bild hundertmal am
 * Tag, über Mobilfunk, auf Kosten des Kunden.
 *
 * DASSELBE GIBT ES AUF iOS (`Verzeichnis.swift`), mit denselben Regeln
 * und denselben Namen.
 */
object Verzeichnis {
    private val faden = Executors.newSingleThreadExecutor()
    private val hauptfaden = Handler(Looper.getMainLooper())

    @Volatile var kollegen: List<Kollege> = emptyList()
        private set
    @Volatile var kontakte: List<Kontakt> = emptyList()
        private set

    private val bilder = HashMap<String, Bitmap>()
    private var zuletzt = 0L

    /** Wird gerufen, wenn neue Daten da sind – die Liste zeichnet neu. */
    var beiAenderung: (() -> Unit)? = null

    /**
     * Nachladen – höchstens alle fünf Minuten, außer jemand besteht darauf.
     *
     * Ein Adressbuch ändert sich selten; die App startet aber oft. Ohne
     * diese Bremse fragte sie bei jedem Wechsel in den Bereich neu.
     */
    fun lade(context: Context, erzwingen: Boolean = false) {
        if (!Dienst.angemeldet(context)) return
        val jetzt = System.currentTimeMillis()
        if (!erzwingen && jetzt - zuletzt < 300_000) return
        zuletzt = jetzt
        val app = context.applicationContext
        faden.execute {
            val leute = Dienst.kollegen(app)
            val buch = Dienst.kontakte(app)
            /* NUR ERSETZEN, WENN ETWAS KAM. Eine leere Antwort nach einem
               Netzfehler würde sonst das Verzeichnis leeren – und beim
               nächsten Anruf stünde wieder eine nackte Nummer da, obwohl
               die Daten längst im Gerät waren. */
            if (leute.isNotEmpty()) kollegen = leute
            if (buch.isNotEmpty()) kontakte = buch
            for (k in leute) if (k.foto.isNotEmpty()) ladeBild(app, k)
            hauptfaden.post { beiAenderung?.invoke() }
        }
    }

    /** Alles vergessen – beim Abmelden. Auch die Bilder auf der Platte. */
    fun leere(context: Context) {
        kollegen = emptyList()
        kontakte = emptyList()
        synchronized(bilder) { bilder.clear() }
        zuletzt = 0L
        ordner(context).deleteRecursively()
        hauptfaden.post { beiAenderung?.invoke() }
    }

    // ---- Rückwärtssuche ----

    /**
     * Wer bin ich?
     *
     * DIE FRAGE IST NICHT TRIVIAL, und deshalb steht sie hier und nicht
     * an drei Stellen: Die App kennt ihren SIP-Benutzernamen, aber der
     * steht in keiner Kollegenliste – und seit die Namen unerratbar
     * sind (`nst-e492zth5a84g`), sagt er auch nichts mehr über den
     * Menschen. Die Anlage nennt beim Tausch des Tokens die interne
     * Nummer der Nebenstelle, und die steht bei genau einem Menschen
     * unter `nebenstellen`.
     *
     * Findet sich keiner, gehört dieser Apparat niemandem (ein
     * Konferenzraum, ein Türsprecher). Dann gibt es kein eigenes Bild,
     * und das ist keine Störung.
     *
     * Dieselbe Regel wie in der iPhone-App (Verzeichnis.ich).
     */
    fun ich(context: Context): Kollege? {
        val meine = Dienst.eigeneNebenstelle(context)
        if (meine.isEmpty()) return null
        return kollegen.firstOrNull { it.nebenstellen.contains(meine) }
    }

    /**
     * Wer ist das? Liefert Namen und – wenn es einen gibt – die Kennung
     * des Menschen, an der sein Bild hängt.
     *
     * DIE REIHENFOLGE IST DIE AUSSAGE: erst die eigenen Kollegen, dann
     * das Adressbuch. Wer intern anruft, ist ein Kollege; stünde
     * dieselbe Nummer zufällig auch im Adressbuch, wäre der Kollege die
     * bessere Auskunft – er hat ein Gesicht.
     */
    fun wer(nummer: String): Pair<String, String?>? {
        val gesucht = normalisiere(nummer)
        if (gesucht.isEmpty()) return null
        for (k in kollegen) {
            if (k.nebenstellen.contains(nummer)) return k.name to k.id
            if (k.durchwahl.isNotEmpty() && k.durchwahl == nummer) return k.name to k.id
            if (k.nummer.isNotEmpty() && normalisiere(k.nummer) == gesucht) return k.name to k.id
        }
        for (e in kontakte) {
            for (n in e.nummern) {
                if (normalisiere(n.nummer) == gesucht) {
                    val wer = if (e.firma.isEmpty()) e.name else "${e.name} · ${e.firma}"
                    return wer to null
                }
            }
        }
        return null
    }

    /**
     * Zwei Nummern vergleichbar machen.
     *
     * +49, 0049 und die führende 0 treffen dasselbe. Alles, was keine
     * Ziffer ist, fliegt raus – Leerzeichen, Bindestriche, Klammern,
     * Schrägstriche schreibt jeder anders.
     */
    fun normalisiere(roh: String): String {
        var z = roh.filter { it.isDigit() || it == '+' }
        if (z.startsWith("+")) z = "00" + z.drop(1)
        z = z.filter { it.isDigit() }
        if (z.startsWith("0049")) z = "0" + z.drop(4)
        return z
    }

    // ---- Bilder ----

    private fun ordner(context: Context) = File(context.cacheDir, "gesichter")

    /** Das Bild eines Menschen, falls schon geladen. */
    fun bild(benutzerId: String?): Bitmap? {
        if (benutzerId == null) return null
        synchronized(bilder) { return bilder[benutzerId] }
    }

    private fun ladeBild(context: Context, k: Kollege) {
        val ziel = File(ordner(context), "${k.id}-${k.foto}.jpg")
        /* SCHON DA? Dann nichts holen. Der Dateiname trägt die Marke –
           ändert sich das Bild, ändert sich der Name, und der alte
           Eintrag wird nie wieder gefunden. */
        if (ziel.exists()) {
            BitmapFactory.decodeFile(ziel.absolutePath)?.let {
                synchronized(bilder) { bilder[k.id] = it }
            }
            return
        }
        val daten = Dienst.foto(context, k.id) ?: return
        val bild = BitmapFactory.decodeByteArray(daten, 0, daten.size) ?: return
        ordner(context).mkdirs()
        ziel.writeBytes(daten)
        synchronized(bilder) { bilder[k.id] = bild }
    }

    /**
     * Zwei Buchstaben aus einem Namen – oder ein Punkt.
     *
     * OHNE BILD BLEIBEN DIE INITIALEN. Ein leerer grauer Kreis sieht aus
     * wie ein Bild, das nicht geladen hat; zwei Buchstaben sehen aus wie
     * eine Entscheidung.
     */
    fun kuerzel(name: String): String {
        val teile = name.split(' ', '.', '-', '_', '·').filter { it.isNotEmpty() }
        return when {
            teile.size >= 2 -> "${teile[0].first()}${teile[1].first()}".uppercase()
            teile.size == 1 -> teile[0].take(2).uppercase()
            else -> "·"
        }
    }
}

File: pruefstuecke/verzeichnis.mjs
/**
 * Prüfstück: Wer ruft da an – und wie sieht er aus?
 *
 * DER AUFTRAG (Inhaber, 05.09.2026): „Ich möchte, dass man selber Fotos
 * hochladen kann … und wenn man sich gegenseitig anruft, dass das Foto
 * von dem Anruf angezeigt wird." Dazu der Satz davor: „Sobald die App
 * ausgeht, muss dann nur nach Benutzername und nach Passwort gefragt
 * werden."
 *
 * DAS GEGENSTÜCK AUF iOS HEISST GENAUSO und prüft dasselbe. Zwei Apps,
 * die verschieden mit der Anlage reden, erzeugen zwei Fehlerbilder für
 * dieselbe Ursache – deshalb stehen hier dieselben Regeln.
 *
 * DIE NORMALISIERUNG WIRD WIRKLICH GERECHNET, nicht nur im Quelltext
 * gesucht: „+49, 0049 und 0 treffen dasselbe" ist die eine Stelle, an
 * der ein Fehler dazu führt, dass beim Anruf der eigenen Kollegin eine
 * nackte Nummer steht.
 *
 * Aufruf:  node pruefstuecke/verzeichnis.mjs
 */
import { readFileSync, existsSync } from "node:fs";

const ergebnisse = [];
const pruefe = (name, ok, extra = "") => {
  ergebnisse.push([name, ok]);
  console.log((ok ? "  ok   " : " FEHL  ") + name + (ok || !extra ? "" : "  – " + extra));
};

const QUELLE = "app/src/main/java/de/dariatech/softphone";
const da = (d) => existsSync(`${QUELLE}/${d}`);
const lies = (d) => (da(d) ? readFileSync(`${QUELLE}/${d}`, "utf-8") : "");
const ohneKommentar = (t) =>
  t.replace(/\/\*[\s\S]*?\*\//g, "").replace(/\/\/[^\n]*/g, "");

const dienst = lies("Dienst.kt");
const verzeichnis = lies("Verzeichnis.kt");
const haupt = ohneKommentar(lies("MainActivity.kt"));
const speicher = lies("Zugangsspeicher.kt");

console.log("\n1) Der Ausweis für den Dienst");
{
  pruefe("es gibt einen Draht zur Anlage", da("Dienst.kt"));
  pruefe("er tauscht Name und Passwort gegen ein Token", /"\/token"/.test(dienst));
  /* IN DER VERSCHLÜSSELTEN ABLAGE, nicht in den gewöhnlichen
     Einstellungen: Wer das Token hat, kommt an das Adressbuch des
     Kunden und an die Bilder seiner Leute. */
  pruefe("das Token liegt in der verschlüsselten Ablage",
    /Zugangsspeicher\.setze\(context, TOKEN/.test(dienst));
  pruefe("und der Speicher kann das auch",
    /fun setze\(context: Context, feld: String/.test(speicher));
  pruefe("beim Abmelden geht es mit", /fun vergiss\(/.test(dienst));
  pruefe("geholt wird es beim Anmelden", /holeAusweis\(\)/.test(haupt));
  /* NIE IM HAUPTFADEN. Android beendet eine App, die dort ins Netz
     geht (NetworkOnMainThreadException) – und zwar sofort. */
  pruefe("aber nie im Hauptfaden", /Thread \{/.test(haupt),
    "Android beendet eine App, die im Hauptfaden ins Netz geht");
}

console.log("\n2) Die Anlage bestimmt keine zweite Adresse");
{
  const anlage = lies("Anlage.kt");
  pruefe("die Adresse des Dienstes steht in Anlage.kt", /const val DIENST/.test(anlage));
  pruefe("und leitet sich vom SIP-Namen ab", /\$SERVER/.test(anlage),
    "wer umzieht, ändert einen DNS-Eintrag und nicht zwei Zeilen in einer App");
  pruefe("sie geht über TLS", /https:\/\//.test(anlage),
    "ein Bearer-Token im Klartext ist ein verschenktes Geheimnis");
}

console.log("\n3) Die Rückwärtssuche – gerechnet, nicht behauptet");
{
  const quelle = verzeichnis.slice(
    verzeichnis.indexOf("fun normalisiere"),
    verzeichnis.indexOf("// ---- Bilder ----")
  );
  const js = quelle
    .replace(/fun normalisiere\(roh: String\): String \{/, "function n(roh) {")
    .replace(/var z = roh\.filter \{ it\.isDigit\(\) \|\| it == '\+' \}/,
      'let z = [...roh].filter((c) => /[0-9]/.test(c) || c === "+").join("")')
    .replace(/if \(z\.startsWith\("\+"\)\) z = "00" \+ z\.drop\(1\)/,
      'if (z.startsWith("+")) z = "00" + z.slice(1);')
    .replace(/z = z\.filter \{ it\.isDigit\(\) \}/,
      'z = [...z].filter((c) => /[0-9]/.test(c)).join("");')
    .replace(/if \(z\.startsWith\("0049"\)\) z = "0" \+ z\.drop\(4\)/,
      'if (z.startsWith("0049")) z = "0" + z.slice(4);')
    .replace(/return z\s*\}/, "return z; }");
  let n = null;
  try {
    n = new Function(`${js}; return n;`)();
  } catch (e) {
    pruefe("die Regel lässt sich nachrechnen", false, String(e));
  }
  if (n) {
    for (const [a, b] of [
      ["+49 831 555 12", "0049831 555-12"],
      ["+49 831 555 12", "0831/55512"],
      ["0049 831 55512", "0831 555 12"]
    ]) {
      pruefe(`„${a}“ und „${b}“ treffen dasselbe`, n(a) === n(b), `${n(a)} vs. ${n(b)}`);
    }
    pruefe("zwei verschiedene Nummern nicht", n("0831 55512") !== n("0831 55513"));
    pruefe("Klammern und Schrägstriche fallen weg",
      n("(0831) 555-12/0") === n("0831555120"));
    pruefe("eine leere Eingabe bleibt leer", n("") === "");
  }
  pruefe("gesucht wird im Gerät", /fun wer\(nummer: String\)/.test(verzeichnis));
  pruefe("die Kollegen zuerst",
    verzeichnis.indexOf("for (k in kollegen)") < verzeichnis.indexOf("for (e in kontakte)"));
}

console.log("\n4) Beim Anruf steht der Name aus dem Verzeichnis");
{
  pruefe("der Anrufer wird nachgeschlagen", /Verzeichnis\.wer\(nummer\)/.test(haupt));
  pruefe("die Kennung für das Bild wird mitgeführt", /gegenueberId = treffer\?\.second/.test(haupt));
  pruefe("und das Gesicht gezeigt", /fun zeigeGesicht\(\)/.test(haupt));
  /* NUR WENN ES EINES GIBT. Ein leerer Kreis sieht aus wie ein Bild,
     das nicht geladen hat. */
  pruefe("ohne Bild bleibt der Platz leer, nicht grau",
    /callerFoto\.visibility = View\.GONE/.test(haupt));
  const layout = existsSync("app/src/main/res/layout/activity_main.xml")
    ? readFileSync("app/src/main/res/layout/activity_main.xml", "utf-8")
    : "";
  pruefe("im Anrufbereich ist Platz dafür", /@\+id\/callerFoto/.test(layout));
}

console.log("\n5) Der Bereich „Kontakte“ ist keine leere Seite mehr");
{
  pruefe("es gibt eine Liste", da("KontakteAdapter.kt"));
  const adapter = lies("KontakteAdapter.kt");
  pruefe("mit zwei Abschnitten", /"Kollegen"/.test(adapter) && /"Adressbuch"/.test(adapter));
  /* SEINE DURCHWAHL, WENN ER EINE HAT – dann klingeln ALLE seine
     Geräte. Wer das umdreht, erwischt den Menschen nur, wenn er
     zufällig am richtigen Apparat sitzt. */
  pruefe("ein Kollege wird über seine Durchwahl gerufen",
    /k\.durchwahl\.isNotEmpty\(\) -> k\.durchwahl/.test(adapter));
  pruefe("und sonst über seine erste Nebenstelle",
    /k\.nebenstellen\.first\(\)/.test(adapter));
  pruefe("ein Druck wählt", /beimAnruf\(z\.ziel\)/.test(adapter));
  pruefe("Bild oder Initialen – immer genau eines",
    /kuerzel\.visibility = View\.INVISIBLE/.test(adapter)
      && /Verzeichnis\.kuerzel\(z\.name\)/.test(adapter));
  pruefe("die Liste hängt im Bereich", /kontakteListe\.adapter = kontakteAdapter/.test(haupt));
  pruefe("und der leere Zustand wechselt mit",
    /kontakteLeer\.visibility/.test(haupt));
}

console.log("\n6) Ein Netzfehler leert das Verzeichnis nicht");
{
  pruefe("nur ersetzen, wenn etwas kam",
    /if \(leute\.isNotEmpty\(\)\) kollegen = leute/.test(verzeichnis),
    "sonst stünde nach einem Netzfehler wieder eine nackte Nummer da");
  pruefe("Bilder tragen ihre Marke im Dateinamen",
    /\$\{k\.id\}-\$\{k\.foto\}\.jpg/.test(verzeichnis));
  pruefe("liegt es schon da, wird nichts geholt", /if \(ziel\.exists\(\)\)/.test(verzeichnis));
}

console.log("\n7) Die Anführungszeichen-Falle");
{
  const suender = [];
  for (const d of ["Dienst.kt", "Verzeichnis.kt", "KontakteAdapter.kt", "MainActivity.kt", "Anlage.kt"]) {
    ohneKommentar(lies(d)).split("\n").forEach((z, i) => {
      for (const m of z.matchAll(/"([^"\\]|\\.)*"/g)) {
        if (/[„“]/.test(m[0])) suender.push(`${d}:${i + 1} ${m[0]}`);
      }
    });
  }
  pruefe("kein deutsches Anführungszeichen in einer Zeichenkette",
    suender.length === 0, suender.join(", "));
}

const fehler = ergebnisse.filter(([, ok]) => !ok);
console.log(`\n${ergebnisse.length - fehler.length}/${ergebnisse.length} Prüfungen bestanden`);
if (fehler.length > 0) {
  console.log("FEHLGESCHLAGEN: " + fehler.map(([n]) => n).join(", "));
  process.exit(1);
}
